File: Example/UIViewControllerTransitionsExample/DemoViewController.m
//
//  DemoViewController.m
//  ModalTransitionAnimator
//
//  Created by Steve Kim on 05/12/2016.
//  Copyright (c) 2016 Steve Kim. All rights reserved.
//

#import "DemoViewController.h"
#import "DragDropTransitionFirstViewController.h"
#import "MoveTransitionFirstViewController.h"
#import "FadeTransitionFirstViewController.h"

@interface DemoViewController ()

@end

@implementation DemoViewController
{
    NSArray<NSString *> *exampleTitles;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"UIViewController Transitions Demo";
    exampleTitles = @[@"UIViewController DragDrop Transition", @"UIViewController Move Transition", @"UIViewController Fade Transition"];
}

// ================================================================================================
//  Protocol Implementation
// ================================================================================================

#pragma mark - UITableView data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return exampleTitles.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 70;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *const cellIdentifier = @"UITableViewCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    
    cell.layer.shouldRasterize = YES;
    cell.layer.rasterizationScale = [UIScreen mainScreen].scale;
    cell.textLabel.text = exampleTitles[indexPath.row];
    
    return cell;
}

#pragma mark - UITableView delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    UIViewController *controller;
    
    if (indexPath.row == 0) {
        controller = [[DragDropTransitionFirstViewController alloc] initWithNibName:@"DragDropTransitionFirstView" bundle:[NSBundle mainBundle]];
    } else if (indexPath.row == 1) {
        controller = [[MoveTransitionFirstViewController alloc] initWithNibName:@"MoveTransitionFirstView" bundle:[NSBundle mainBundle]];
    } else if (indexPath.row == 2) {
        controller = [[FadeTransitionFirstViewController alloc] initWithNibName:@"FadeTransitionFirstView" bundle:[NSBundle mainBundle]];
    }
    
    [self.navigationController pushViewController:controller animated:YES];
}

@end

File: Example/UIViewControllerTransitionsExample/Source/DemoViewController.swift
//  BSD 2-Clause License
//
//  Copyright (c) 2016 ~ 2021, Steve Kim
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//  DemoViewController.swift
//  UIViewControllerTransitionsExample
//
//  Created by pisces on 14/04/2017.
//  Copyright © 2017 Steve Kim. All rights reserved.
//

import UIViewControllerTransitions

final class DemoViewController: UITableViewController {
    
    private let titles: [String] = [
        "DragDrop",
        "Move",
        "Fade",
        "Move for Navigation",
        "Zoom"
    ]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        clearsSelectionOnViewWillAppear = true
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return titles.count
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 70
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let reuseIdentifier = "UITableViewCell"
        var cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier)
        
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: reuseIdentifier)
            cell!.accessoryType = .disclosureIndicator
        }
        
        return cell!
    }
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        cell.textLabel?.text = titles[indexPath.row]
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        var viewController: UIViewController?
        
        switch indexPath.row {
        case 0:
            viewController = DragDropTransitionFirstViewController(nibName: "DragDropTransitionFirstView", bundle: .main)
        case 1:
            viewController = MoveTransitionFirstViewController(nibName: "MoveTransitionFirstView", bundle: .main)
        case 2:
            viewController = FadeTransitionFirstViewController(nibName: "FadeTransitionFirstView", bundle: .main)
        case 3:
            viewController = NavigationMoveTransitionFirstViewController(nibName: "NavigationMoveTransitionFirstView", bundle: .main)
        case 4:
            viewController = UIStoryboard(name: "ZoomTransition", bundle: nil).instantiateViewController(withIdentifier: "FirstScene")
        default:
            break
        }
        
        guard let controller = viewController else {return}
        
        if indexPath.row == 3 {
            let navigationController = UINavigationController(rootViewController: controller)
            navigationController.modalPresentationStyle = .fullScreen
            present(navigationController, animated: true, completion: nil)
        } else {
            navigationController?.pushViewController(controller, animated: true)
        }
    }
}
